### PR TITLE
feat: add jwt-auth authorization method

### PR DESCRIPTION
**Pull request by AlinsRan**
### Type of change:

- [ ] Bugfix
- [x] New feature provided
- [ ] Improve performance
- [ ] Backport patches

### What this PR does / why we need it:

Issues: #1001 

### Pre-submission checklist:

* [ ] Did you explain what problem does this PR solve? Or what new features have been added?
* [x] Have you added corresponding test cases?
* [x] Have you modified the corresponding document?
* [ ] Is this PR backward compatible? **If it is not backward compatible, please discuss on the [mailing list](https://github.com/apache/apisix-ingress-controller#community) first**

### CODE REVIEW

File: pkg/kube/apisix/apis/config/v2/types.go [hunk 1]
```diff
@@ -355,6 +365,23 @@ type ApisixConsumerKeyAuthValue struct {
...
+}
+
+// ApisixConsumerJwtAuthValue defines the in-place configuration for jwt auth.
+type ApisixConsumerJwtAuthValue struct {
+	Key          string `json:"key" yaml:"key"`
+	Secret       string `json:"secret,omitempty" yaml:"secret,omitempty"`
+	PublicKey    string `json:"public_key,omitempty" yaml:"public_key,omitempty"`
+	PrivateKey   string `json:"private_key" yaml:"private_key,omitempty"`
+	Algorithm    string `json:"algorithm,omitempty" yaml:"algorithm,omitempty"`
+	Exp          int64  `json:"exp,omitempty" yaml:"exp,omitempty"`
```
Comment (tao12345666333) [inline pkg/kube/apisix/apis/config/v2/types.go:381]: exp is a positive number, I think you should add a check

Comment (tao12345666333) [reply pkg/kube/apisix/apis/config/v2/types.go:381]: exp is a positive number, I think you should add a check

**Review by tao12345666333 — COMMENTED** [review]
Need to add a checksum, other contents LGTM

File: pkg/kube/apisix/apis/config/v2/types.go (hunk 1 above)
Comment (AlinsRan) [reply pkg/kube/apisix/apis/config/v2/types.go:381]: Solved!

File: pkg/kube/translation/plugin.go [hunk 2]
```diff
@@ -115,6 +116,9 @@ func (t *translator) translateConsumerBasicAuthPlugin(consumerNamespace string,
 
 func (t *translator) translateConsumerJwtAuthPlugin(consumerNamespace string, cfg *configv2beta3.ApisixConsumerJwtAuth) (*apisixv1.JwtAuthConsumerConfig, error) {
 	if cfg.Value != nil {
+		if cfg.Value.Exp < 0 {
```
Comment (tao12345666333) [inline pkg/kube/translation/plugin.go:119]: https://github.com/apache/apisix/blob/ab7a7675d2a1abb1f0299a8fd44e5f80435c2158/apisix/plugins/jwt-auth.lua#L66

the minimum is 1

File: pkg/kube/translation/plugin.go [hunk 3]
```diff
@@ -141,6 +145,9 @@ func (t *translator) translateConsumerJwtAuthPlugin(consumerNamespace string, cf
 	}
 	expRaw := sec.Data["exp"]
 	exp, _ := strconv.ParseInt(string(expRaw), 10, 64)
+	if exp < 0 {
```
Comment (tao12345666333) [inline pkg/kube/translation/plugin.go:148]: ditto

**Review by tao12345666333 — CHANGES_REQUESTED** [review]

**Comment by tao12345666333** [comment]
ping @sff0219

**Review by tao12345666333 — APPROVED** [review]
LGTM

File: pkg/kube/apisix/apis/config/v2beta2/types.go [hunk 4]
```diff
@@ -140,6 +141,14 @@ type ApisixRouteAuthenticationKeyAuth struct {
 	Header string `json:"header,omitempty" yaml:"header,omitempty"`
 }
 
+// ApisixRouteAuthenticationJwtAuth is the jwtAuth-related
+// configuration in ApisixRouteAuthentication.
+type ApisixRouteAuthenticationJwtAuth struct {
```
Comment (lingsamuel) [inline pkg/kube/apisix/apis/config/v2beta2/types.go:146]: we are going to add new feature to v2beta2? @tao12345666333 is it by design?

Comment (tao12345666333) [reply pkg/kube/apisix/apis/config/v2beta2/types.go:146]: no, we don't need add any feature to v2beta2.  I missed.
 @AlinsRan please remove `v2beta2` related code. we only need add feature to `v2beta3` and `v2`

Comment (AlinsRan) [reply pkg/kube/apisix/apis/config/v2beta2/types.go:146]: OK.

File: pkg/kube/apisix/apis/config/v2/types.go [hunk 5]
```diff
@@ -180,6 +181,14 @@ type ApisixRouteAuthenticationKeyAuth struct {
 	Header string `json:"header,omitempty" yaml:"header,omitempty"`
 }
 
+// ApisixRouteAuthenticationJwtAuth is the jwtAuth-related
```
Comment (tokers) [inline pkg/kube/apisix/apis/config/v2/types.go:184]:
```suggestion
// ApisixRouteAuthenticationJwtAuth is the jwt auth related
```

File: pkg/kube/apisix/apis/config/v2/types.go [hunk 6]
```diff
@@ -330,6 +339,7 @@ type ApisixConsumerSpec struct {
 type ApisixConsumerAuthParameter struct {
 	BasicAuth *ApisixConsumerBasicAuth `json:"basicAuth,omitempty" yaml:"basicAuth"`
 	KeyAuth   *ApisixConsumerKeyAuth   `json:"keyAuth,omitempty" yaml:"keyAuth"`
+	JwtAuht   *ApisixConsumerJwtAuth   `json:"jwtAuth,omitempty" yaml:"jwtAuth"`
```
Comment (tokers) [inline pkg/kube/apisix/apis/config/v2/types.go:342]:
```suggestion
	JwtAuth   *ApisixConsumerJwtAuth   `json:"jwtAuth,omitempty" yaml:"jwtAuth"`
```

Comment (tokers) [inline pkg/kube/apisix/apis/config/v2/types.go:342]:
```suggestion
	JwtAuth   *ApisixConsumerJwtAuth   `json:"jwtAuth,omitempty" yaml:"jwtAuth"`
```

File: pkg/types/apisix/v1/plugin_types.go [hunk 7]
```diff
@@ -70,6 +70,19 @@ type BasicAuthConsumerConfig struct {
 	Password string `json:"password"`
 }
 
+// JwtAuthConsumerConfig is the rule config for basic-auth plugin
```
Comment (tokers) [inline pkg/types/apisix/v1/plugin_types.go:73]:
```suggestion
// JwtAuthConsumerConfig is the rule config for jwt-auth plugin
```

File: samples/deploy/crd/v1/ApisixConsumer.yaml [hunk 8]
```diff
@@ -98,3 +99,37 @@ spec:
...
+                          required:
+                            - key
+                        secretRef:
+                          type: object
+                          properties:
+                            name:
+                              type: string
+                              minLength: 1
+                          required:
+                            - name
```
Comment (tokers) [inline samples/deploy/crd/v1/ApisixConsumer.yaml:135]: Add the EOL symbol.

**Review by tokers — COMMENTED** [review]
Missing docs to:

1. JWT Auth Config Reference
3. How to use it?

File: pkg/kube/apisix/apis/config/v2/types.go (hunk 5 above)
Comment (AlinsRan) [reply pkg/kube/apisix/apis/config/v2/types.go:184]: Thanks.

**Comment by AlinsRan** [comment]
> I don't think we need to write a separate document. Content can be added to this document. https://github.com/apache/apisix-ingress-controller/blob/master/docs/en/latest/practices/enable-authentication-and-restriction.md

I agree.

**Review by tao12345666333 — APPROVED** [review]

**Comment by AlinsRan** [comment]
> Missing docs to:
> 
> 1. JWT Auth Config Reference
> 2. How to use it?

doc: #1018

**Review by lingsamuel — APPROVED** [review]

**Comment by tao12345666333** [comment]
Let's move forward, docs can be tracked with #1018

cc @tokers

**Pull request merged**